src/query/service/src/pipelines/builders/copy.rs

```diff
@@ -117,74 +87,151 @@ pub fn build_append_data_with_finish_pipeline(
+                ctx.clone(),
+                main_pipeline,
+                to_table,
+                plan_required_values_schema,
+                copied_files,
+                false,
+                AppendMode::Copy,
+            )?;
         }
         CopyIntoTableMode::Replace => {}
     }
 
-    if local_id == ctx.get_cluster().local_id {
+    // set finished callback to pipeline.
+    set_pipeline_finish_callback(
+        ctx,
+        main_pipeline,
+        coordinate_node_id,
+        stage_info,
+        files,
+        purge,
+    )
+}
+
+/// Build a pipeline for append data for distributed mode.
+/// 1. only append data to table not commit
+/// 2. not purge files
+pub fn build_distributed_append_data_pipeline(
+    ctx: Arc<QueryContext>,
+    main_pipeline: &mut Pipeline,
+    plan: DistributedCopyIntoTable,
+    source_schema: Arc<DataSchema>,
+    to_table: Arc<dyn Table>,
+) -> Result<()> {
+    let plan_required_source_schema = plan.required_source_schema;
+    let plan_required_values_schema = plan.required_values_schema;
+    let plan_values_consts = plan.values_consts;
+    let plan_stage_table_info = plan.stage_table_info;
+    let plan_write_mode = plan.write_mode;
+    let coordinator_node_id = plan.local_node_id;
+
+    if source_schema != plan_required_source_schema {
+        // only parquet need cast
+        let func_ctx = ctx.get_function_context()?;
+        main_pipeline.add_transform(|transform_input_port, transform_output_port| {
+            TransformCastSchema::try_create(
+                transform_input_port,
+                transform_output_port,
+                source_schema.clone(),
+                plan_required_source_schema.clone(),
+                func_ctx.clone(),
+            )
+        })?;
+    }
+
+    if !plan_values_consts.is_empty() {
+        fill_const_columns(
+            ctx.clone(),
+            main_pipeline,
+            source_schema,
+            plan_required_values_schema.clone(),
+            plan_values_consts,
+        )?;
+    }
+
+    let stage_info = plan_stage_table_info.stage_info;
+    // Only append to table, not commit.
+    let write_mode = plan_write_mode;
+    match write_mode {
+        CopyIntoTableMode::Insert { overwrite: _ } => build_append2table_without_commit_pipeline(
+            ctx.clone(),
+            main_pipeline,
+            to_table,
+            plan_required_values_schema,
+            AppendMode::Copy,
+        )?,
+        CopyIntoTableMode::Copy => build_append2table_without_commit_pipeline(
+            ctx.clone(),
+            main_pipeline,
+            to_table,
+            plan_required_values_schema,
+            AppendMode::Copy,
+        )?,
+        CopyIntoTableMode::Replace => {}
+    }
+
+    // set finished callback to pipeline.
+    set_pipeline_finish_callback(
+        ctx,
+        main_pipeline,
+        coordinator_node_id,
+        stage_info,
+        vec![],
+        false,
```

Comment:
should we get a purge flag here? not direct false?

Reply:
The distributed can not purge file, the purge will only do in the coordinator node, so here should false.

Reply:
The logic test is good to test my fault. Fixed.